src/cowboy_http2.erl

```diff
@@ -563,14 +576,21 @@ early_error(State0=#state{ref=Ref, opts=Opts, peer=Peer},
 		send_headers(State0, StreamID, fin, StatusCode0, RespHeaders0)
 	end.
 
-rst_stream_frame(State=#state{streams=Streams0, children=Children0}, StreamID, Reason) ->
-	case maps:take(StreamID, Streams0) of
+rst_stream_frame(State0=#state{streams=Streams0, children=Children0}, StreamID, Reason) ->
+	State1 = case maps:take(StreamID, Streams0) of
 		{#stream{state=StreamState}, Streams} ->
-			terminate_stream_handler(State, StreamID, Reason, StreamState),
+			terminate_stream_handler(State0, StreamID, Reason, StreamState),
 			Children = cowboy_children:shutdown(Children0, StreamID),
-			State#state{streams=Streams, children=Children};
+			State0#state{streams=Streams, children=Children};
 		error ->
-			State
+			State0
+	end,
+	case cancel_rate(State1) of
```

Comment:
This should only be applied if a stream was indeed reset. It's possible the stream was already gone (so we didn't waste resources) or it's also possible that Cowboy is seen as misbehaving and the client sends more than one RST_STREAM for a single stream. Since the purpose is to protect against actual stream cancellations (and not frame flood, covered by a separate setting), we should only count those actual cancellations.